tools/wake-format/emitter.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <wcl/optional.h>
+#include <wcl/rope.h>
+
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+#include "parser/cst.h"
+
+class Emitter {
+ public:
+  struct ctx_t {
+    int nest_level = 0;
+    bool is_flat = false;
+
+    ctx_t flat() {
+      ctx_t copy = *this;
+      copy.is_flat = true;
+      return copy;
+    }
+  };
+
+  struct nest_t {
+    ctx_t* ctx;
+
+    nest_t(ctx_t* ctx) : ctx(ctx) { ctx->nest_level++; }
+
+    ~nest_t() { ctx->nest_level--; }
+  };
+
+  Emitter(std::ostream* ostream) : ostream(ostream) { assert(ostream != nullptr); }
+
+  // Walks the CST, formats it, and writes it to ostream
+  void layout(CST cst);
+
+ private:
+  // Top level tree walk. Dispatches out the calls for various nodes
+  wcl::optional<wcl::rope> walk_top(ctx_t ctx, CSTElement node);
+
+  // Returns a nest_t instance that increases the indent level
+  // until the instance goes out of scope
+  nest_t nest(ctx_t ctx);
+
+  // Emits a newline and any indentation needed
+  wcl::rope newline(ctx_t ctx);
+
+  // Emits a space character `count` times
+  wcl::rope space(ctx_t ctx, uint8_t count = 1);
+
+  // TODO: memoize this function
+  // returns the subtree with all newlines and indentation removed
+  // if possible, None otherwise
```

Comment:
minor nit: I think this doesn't care about indentation, its just that indentation follows newlines of which there are none here.

Reply:
I'm not sure I follow, we need to remove the indentation for flat to work no?

Reply:
Oh yeah I guess when a soft newline is emitted you don't emit the indentation true. I guess I think of indentation and newlines are being related things so the entire `newline()` call would just return when the context is falt. But you're right that also means indentation is ignored too.